### ISSUE TITLE
ics calendar entries may end up in a different time zone (re-write #217)

### ISSUE THREAD

**Issue by KlausGPaul** [Attendize/Attendize#292]
It is s slightly complicated set-up: In my setup, attendize runs in the cloud in East Coast as a dockerized container. The event takes place in a Central European time zone, but the network, from where most people access attendize, connects to the internet using an endpoint located in a GMT/BST time zone.

So, when I set up the event as 9:00-10:00, recipients receive an ics that shows 10:00-11:00, it _appears_ as if somewhere in the process, a time zone adjustment takes place. Tickets themselves have the correct time. I was expecting tickets to have the very same time as the ics, or an ability for specifying time zones somewhere (that would be feature request).

None of the machines (cloud or docker) have been configured specifically for any time zone in particular. The phenomenon occurs irrespective of location (the dummy event, see attached, would take place in Georgia, there is an actual Crab Shack there, still it is being adjusted by +1 hr).

Machines used to order are configured to be in the central european time zone.

This issue supercedes 217 was assuming a different scenario, whichturned out to be incorrect. 217 I closed as non-issue.
![dummy event calendar](https://cloud.githubusercontent.com/assets/20340525/22054502/2e07e15e-dd56-11e6-87b7-11e90d3b8475.PNG)
![dummy event page](https://cloud.githubusercontent.com/assets/20340525/22054503/2e1a7fd0-dd56-11e6-81e1-c422ca84a87f.PNG)
![dummy event calendar ios](https://cloud.githubusercontent.com/assets/20340525/22054504/2e68723a-dd56-11e6-9fd7-96ad15e9cdc4.jpeg)

**Comment by jmowatt** [comment]
Can you paste the contents of the ICS file please?

**Comment by KlausGPaul** [comment]
Apologies @jmowatt , I had already shut down the instance. I just removed the IP address of the machine.

> BEGIN:VCALENDAR
VERSION:2.0
PRODID:http://example.com
BEGIN:VEVENT
UID:http://example.com/e/2/dummy-time-zone-test-event
DTSTAMP:20170122T163432Z
DTSTART:20170119T090000Z
DTEND:20170119T100000Z
SUMMARY:Dummy Time Zone Test Event
LOCATION:The Crab Shack
DESCRIPTION:NA
END:VEVENT
END:VCALENDAR

**Comment by KlausGPaul** [comment]
I am speculating now, this could be an email issue. The emails may be sent with an associated time zone in the header which is then interpreted as an event time zone. I also found [this on stackoverflow](http://stackoverflow.com/questions/11055947/does-outlook-correctly-handle-time-zones-for-ics-icalendar-files), so maybe it it required to add time zone(s) to the event [see also](http://stackoverflow.com/questions/23963459/importing-ics-into-google-calendar-with-correct-timezone). So, maybe this turns into a feature request for event (and ics) timezones.

**Comment by KlausGPaul** [comment]
Just experimented with this further.... The current scheme also gets confused with daylight savings time, I created a demo event on April 1, 12:00-13:00, and it got imported into my local outlook (which directly connects to the internet, so it would not suffer from the above) as 14:00-15:00.

> `BEGIN:VCALENDAR`
> `VERSION:2.0`
> `PRODID:https://test.com`
> `BEGIN:VEVENT`
> `UID:https://test.com/e/6/test-and-demo-event`
> `DTSTAMP:20170320T103516Z`
> `DTSTART:20170401T120000Z`
> `DTEND:20170401T130000Z`
> `SUMMARY:Test and Demo Event`
> `LOCATION:Whereever`
> `DESCRIPTION:Just Test and Demo`
> `END:VEVENT`
> `END:VCALENDAR`

![image](https://cloud.githubusercontent.com/assets/20340525/24097849/f51e9db4-0d67-11e7-9c2d-b970b0109b68.png)

This is because the DTSTART and DTEND entries are explicitly marked as Z (zulu) timestamps, which is UTC, which does not do daylight savings.

As a dirty workaround, I have changed `app/Models/Event.php` to generate the timestamps without the trailing Z. This is probably not quite clean as it does not strictly define the timezone anymore, but it appears as if it was working (need to test from within our company network).

> `diff output`
> `< DTSTAMP:{$timestamp->format('Ymd\THis')}`
> `< DTSTART:{$start_date->format('Ymd\THis')}`
> `< DTEND:{$end_date->format('Ymd\THis')}`
> `---`
> `> DTSTAMP:{$timestamp->format('Ymd\THis\Z')}`
> `> DTSTART:{$start_date->format('Ymd\THis\Z')}`
> `> DTEND:{$end_date->format('Ymd\THis\Z')}`
> `

Now the ics file contains non-zulu times, which, at the mercy of the email/calendar application, seems to be the correct time.

> `BEGIN:VCALENDAR`
> `VERSION:2.0`
> `PRODID:https://test.com`
> `BEGIN:VEVENT`
> `UID:https://test.com/e/6/test-and-demo-event`
> `DTSTAMP:20170320T103516`
> `DTSTART:20170401T120000`
> `DTEND:20170401T130000`
> `SUMMARY:Test and Demo Event`
> `LOCATION:Whereever`
> `DESCRIPTION:Just Test and Demo`
> `END:VEVENT`
> `END:VCALENDAR`

![image](https://cloud.githubusercontent.com/assets/20340525/24097812/d2d3d648-0d67-11e7-811f-b25b4a788ba1.png)

**Comment by johannac** [comment]
I don't think this is an issue that needs to be addressed. The ICS file contains time in UTC and it's formatted correctly.

**Comment by KlausGPaul** [comment]
Strange, I had forgotten about this, but, FYI, the conversation contains a bug fix for an effect that actually happened. Why is this a non-issue (unless removing the Z zulu timestamp did happen)?
Just asking...

**Comment by johannac** [comment]
Hi @KlausGPaul I had looked through the ICS spec and some other sites/blogs and it seemed to me that the 'Z' is required when no timezone/TZID or TZOFFSETFROM/TZOFFSETTO is specified. That's my understanding anyway, it's not the easiest thing to research 😝

**Comment by KlausGPaul** [comment]
Hi @johannac no worries --- this issue had caused me a lot of grief in my previous life when we used attendize to organise internal company events. This used a Windows/Outlook (server) setup (but I faintly recall this alos happened on Apple calendar entries). I fixed it by removing the Zulu label from the timestamp (after all, Z is a specific timezone).
Its been a while, not sure of the code base has changed, php is not my sweet spot, and, at the time, the project was a bit dead so I did not go for a pull request.... Maybe the issue has already cleared due to some other change....
Hope you don't mind my comment, but I still recall the emails I received complaining about the wrong caledar entries.
